{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# attempted to refactor Zack's code \n",
    "# ended up writing my own script from scratch \n",
    "# used different implementation for greater efficiency\n",
    "\n",
    "# current stage: finished calling sites\n",
    "# next to-do: finish generate_probe(); output as probe file for JProbe\n",
    "## JProbe probe-generator inputs: peak seqs, kmer, pwm, probe length, binding site, window, escore\n",
    "## TO HANDLE: pwm, window(?)\n",
    "# last updated on 9/18/19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "inputHidden": false,
    "outputHidden": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools # might not be necessary\n",
    "from Bio import SeqIO\n",
    "import os, subprocess, re\n",
    "from Bio.Seq import reverse_complement\n",
    "import pathlib # might not be necessary, will clean up later\n",
    "from itertools import groupby\n",
    "from operator import itemgetter\n",
    "\n",
    "wkdir, TF = \"/mnt/c/Users/th184/Box Sync/Gordan Lab/PRIORITY_score_caller\", \"Elk1\" # run from Ubuntu hence /mnt/c\n",
    "peak_coord_file = \"Elk1DHSv2promoterHg19NoDac.bed\" \n",
    "kmer_file = \"Elk1_kmers_45_PRIORITY.txt\"\n",
    "genome_file = \"/mnt/c/Users/th184/Box Sync/Probe Design/hg19.fa\"\n",
    "\n",
    "kmer_len = 7\n",
    "overlap_req = 2\n",
    "escore_cutoff = 0.45 # maybe 0.40?\n",
    "probe_len = 36\n",
    "# Prior assumptions about the kmer PRIORITY alignment\n",
    "core = [7,12]  # 0-indexed\n",
    "center_pos = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "inputHidden": false,
    "outputHidden": false
   },
   "outputs": [],
   "source": [
    "io_folder = pathlib.Path(wkdir) / TF\n",
    "os.chdir(io_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fill_space(path):\n",
    "    s = pathlib.PurePath(path).split()\n",
    "    print(s)\n",
    "    return \"\\ \".join((path).split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read CSV, filter for short seqs and duplicates, adjust coordinates\n",
    "pk_coord = pd.read_csv(io_folder/peak_coord_file, sep = '\\t', header = None, usecols=[0,1,2]) \n",
    "pk_coord = pk_coord[pk_coord[2]-pk_coord[1] > (kmer_len+overlap_req)].drop_duplicates() # filter out(?) short sequences the caller would have trouble with\n",
    "pk_coord[1] = pk_coord[1] - 1 # Move back 1 to adjust for bedtools getfasta (0-indexed)\n",
    "\n",
    "pk_coord_adj_file = peak_coord_file.split(\".\")[0] + \"_0-indexed.bed\"\n",
    "pk_file = peak_coord_file.split(\".\")[0]+\".fasta\"\n",
    "# need to figure the space issue out\n",
    "pk_coord.to_csv(pk_coord_adj_file, sep = '\\t', index = False, header = None) # output to getfasta\n",
    "# subprocess.call([\"bedtools\", \"getfasta\", \"-s\", \"-fi\", fill_space(genome_file), \"-bed\", fill_space(io_folder)/pk_coord_adj_file, \">\", fill_space(io_folder)/pk_file])\n",
    "# subprocess.call(f'bedtools getfasta -s -fi {genome_file} -bed {io_folder/pk_coord_adj_file} > {io_folder/pk_file}')\n",
    "# subprocess.call(f'bedtools getfasta -s -fi {fill_space(genome_file)} -bed {fill_space(pk_coord_adj_file)} > {fill_space(pk_file)}')\n",
    "# subprocess.call(r\"bedtools getfasta -s -fi /mnt/c/Users/th184/Box\\ Sync/Probe\\ Design/hg19.fa -bed /mnt/c/Users/th184/Box\\ Sync/Gordan\\ Lab/PRIORITY_score_caller/Elk1/Elk1DHSv2promoterHg19NoDac_0-indexed.bed > /mnt/c/Users/th184/Box\\ Sync/Gordan\\ Lab/PRIORITY_score_caller/Elk1/Elk1DHSv2promoterHg19NoDac.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "inputHidden": false,
    "outputHidden": false
   },
   "outputs": [],
   "source": [
    "def get_kmer(kmer_file, kmer_len):\n",
    "    # return kmer_list as a Set, kmer_core_id as a dictionary\n",
    "    kmers = pd.read_csv(io_folder/kmer_file , sep = '\\t') # created an Unnamed column for no reason\n",
    "    kmers = kmers.loc[:, ~kmers.columns.str.contains('^Unnamed')] # get rid of the Unnamed column\n",
    "    kmer_set = set()\n",
    "    kmer_core_id = {} # {kmer: boolean indicating whether is core}\n",
    "    for i in kmers['sequences']:\n",
    "        kPos = re.search('(C|G|A|T)',i).start() # 0-indexed\n",
    "        kmer = i[kPos: kPos+kmer_len]\n",
    "        kmer_set.add(kmer)\n",
    "        kmer_core_id[kmer] = False if ('.' in i[core[0]:core[1]+1]) else True\n",
    "    return kmer_set, kmer_core_id\n",
    "\n",
    "def read_fasta(file):\n",
    "    entry = []\n",
    "    with open(pk_file_test, \"r\") as input_handle:\n",
    "        for record in SeqIO.parse(input_handle, \"fasta\"):\n",
    "            entry.append([record.id, str(record.seq).upper(), len(record.seq), str(reverse_complement(record.seq)).upper()])\n",
    "    arr = np.array(entry)\n",
    "    df = pd.DataFrame({'seq_id':arr[:, 0], 'seq':arr[:,1], 'seq_len':arr[:,2], 'rev_comp':arr[:,3]})\n",
    "    return df\n",
    "\n",
    "def get_pos(seq, kmer_len, kmer_set, kmer_core_id_dict):\n",
    "    # return a list of pos at which kmer exists in the seq\n",
    "    pos, core_id = [], []\n",
    "    for i in range(len(seq) - kmer_len + 1):\n",
    "        window = seq[i:i+kmer_len]\n",
    "        if window in kmer_set:\n",
    "            pos.append(i)\n",
    "            core_id.append(kmer_core_id_dict[window])\n",
    "    return pd.Series([pos, core_id])\n",
    "\n",
    "def group_by_tuple(a, b):\n",
    "    '''\n",
    "    a = [(1,2), (3,), (4, 5, 6)]\n",
    "    b = [True, True, False, True, False, True]\n",
    "    output = [(True, True), (False,), (True, False, True)]\n",
    "    '''\n",
    "    prev_tup_head, curr_tup_head = 0, 0\n",
    "    grouped_b = []\n",
    "    for tup in a:\n",
    "        curr_tup_head = prev_tup_head+len(tup)\n",
    "        boo = tuple(b[prev_tup_head:curr_tup_head])\n",
    "        grouped_b.append(boo)\n",
    "        prev_tup_head = curr_tup_head\n",
    "    return grouped_b\n",
    "        \n",
    "def call_sites(pos_list, kmer_core_bool, overlap_req):\n",
    "    # binding site : (1) len(kmer_pos) >= overlap_req (2) contain at least one core_kmer\n",
    "    # one site = one tuple\n",
    "    grouped_pos, TFBS = [], []\n",
    "    # get runs of consecutive numbers using groupby \n",
    "    for k, g in groupby(enumerate(pos_list), lambda x: x[1] - x[0]):\n",
    "        grouped_pos.append(tuple(map(itemgetter(1), g)))\n",
    "    grouped_core_bool = group_by_tuple(grouped_pos, kmer_core_bool)\n",
    "    for pos_tuple, core_bool_tuple in zip(grouped_pos, grouped_core_bool):\n",
    "        if len(pos_tuple) >= overlap_req and any(core_bool_tuple):\n",
    "            TFBS.append(pos_tuple)\n",
    "    return TFBS\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>rev_comp</th>\n",
       "      <th>kmer_pos</th>\n",
       "      <th>kmer_core_bool</th>\n",
       "      <th>TF_Binding_Sites</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:39491846-39491989()</td>\n",
       "      <td>CCTCCCTCAGCCTTCCAAGGTGCTGGGATTGCAGACGTCAGCCACC...</td>\n",
       "      <td>143</td>\n",
       "      <td>GACCTCAGCGAGACGTTTGCAGCGCGACGCCGGAAGCCGACCTTCC...</td>\n",
       "      <td>[106, 107, 108, 109]</td>\n",
       "      <td>[False, True, True, False]</td>\n",
       "      <td>[(106, 107, 108, 109)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11:46722049-46722101()</td>\n",
       "      <td>CCTTTCCTTTCCCCATCCCCACCCGGGGAGACCCCGCGCCCAGCAG...</td>\n",
       "      <td>52</td>\n",
       "      <td>GGTGAGCTGCTGGGCGCGGGGTCTCCCCGGGTGGGGATGGGGAAAG...</td>\n",
       "      <td>[22]</td>\n",
       "      <td>[False]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11:46722148-46722368()</td>\n",
       "      <td>CATCCGGCTCCGCGGTGGAACGGGAACCGCCCCCTCCCGCTCTCTT...</td>\n",
       "      <td>220</td>\n",
       "      <td>CGGAACTAATTGTTCCAGCCATCTCTTTGGGAGCGACGAGTCGGAC...</td>\n",
       "      <td>[0, 1, 2, 58, 59, 60, 79, 80, 81, 82, 83, 119,...</td>\n",
       "      <td>[True, True, False, False, True, True, False, ...</td>\n",
       "      <td>[(0, 1, 2), (58, 59, 60), (79, 80, 81, 82, 83)]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      seq_id  \\\n",
       "0   chr1:39491846-39491989()   \n",
       "1  chr11:46722049-46722101()   \n",
       "2  chr11:46722148-46722368()   \n",
       "\n",
       "                                                 seq seq_len  \\\n",
       "0  CCTCCCTCAGCCTTCCAAGGTGCTGGGATTGCAGACGTCAGCCACC...     143   \n",
       "1  CCTTTCCTTTCCCCATCCCCACCCGGGGAGACCCCGCGCCCAGCAG...      52   \n",
       "2  CATCCGGCTCCGCGGTGGAACGGGAACCGCCCCCTCCCGCTCTCTT...     220   \n",
       "\n",
       "                                            rev_comp  \\\n",
       "0  GACCTCAGCGAGACGTTTGCAGCGCGACGCCGGAAGCCGACCTTCC...   \n",
       "1  GGTGAGCTGCTGGGCGCGGGGTCTCCCCGGGTGGGGATGGGGAAAG...   \n",
       "2  CGGAACTAATTGTTCCAGCCATCTCTTTGGGAGCGACGAGTCGGAC...   \n",
       "\n",
       "                                            kmer_pos  \\\n",
       "0                               [106, 107, 108, 109]   \n",
       "1                                               [22]   \n",
       "2  [0, 1, 2, 58, 59, 60, 79, 80, 81, 82, 83, 119,...   \n",
       "\n",
       "                                      kmer_core_bool  \\\n",
       "0                         [False, True, True, False]   \n",
       "1                                            [False]   \n",
       "2  [True, True, False, False, True, True, False, ...   \n",
       "\n",
       "                                  TF_Binding_Sites  \n",
       "0                           [(106, 107, 108, 109)]  \n",
       "1                                               []  \n",
       "2  [(0, 1, 2), (58, 59, 60), (79, 80, 81, 82, 83)]  "
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmer_set, kmer_core_id_dict = get_kmer(kmer_file, kmer_len)\n",
    "pk_file_test = \"Elk1DHSv2promoterHg19NoDac_test.fasta\"\n",
    "peak_df = read_fasta(pk_file_test)\n",
    "peak_df[[\"kmer_pos\",\"kmer_core_bool\"]] = peak_df.apply(lambda peak_df: get_pos(peak_df[\"seq\"], kmer_len, kmer_set, kmer_core_id_dict), axis = 1)\n",
    "peak_df[\"TF_Binding_Sites\"] = peak_df.apply(lambda peak_df: call_sites(peak_df[\"kmer_pos\"], peak_df[\"kmer_core_bool\"], overlap_req), axis = 1)\n",
    "peak_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_probes(kmer_pos_lst, seq, probe_len):\n",
    "    # input: kmer_pos is a list of kmer positions\n",
    "    # generate one probe for each binding site by centering it and trimming to probe_len (no mutation is done)\n",
    "    probe_half_len = probe_len//2 # floor division\n",
    "    BS_middle_index = (len(kmer_pos_lst) - 1)//2 \n",
    "    BS_middle_pos = kmer_pos_lst[BS_middle_index]\n",
    "    head, tail = 0, 0\n",
    "    if probe_len%2==0:\n",
    "        head = BS_middle_pos - probe_half_len + 1\n",
    "        tail = BS_middle_pos + probe_half_len\n",
    "    elif len(kmer_pos_lst)%2==0: # odd probe_len, even kmer_pos_lst\n",
    "        head = BS_middle_pos - probe_half_len + 1\n",
    "        tail = BS_middle_pos + probe_half_len + 1\n",
    "    else: # odd probe_len, odd kmer_pos_lst\n",
    "        head = BS_middle_pos - probe_half_len\n",
    "        tail = BS_middle_pos + probe_half_len\n",
    "    probe = seq[head: tail+1]\n",
    "    return probe\n",
    "\n",
    "peak_df[\"probe\"]\n",
    "    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_kmer_OLD(kmer_file):\n",
    "    # no need to put things into dataframe\n",
    "    kmers = pd.read_csv(io_folder/kmer_file , sep = '\\t') # created an Unnamed column for no reason\n",
    "    kmers = kmers.loc[:, ~kmers.columns.str.contains('^Unnamed')] # get rid of the Unnamed column\n",
    "    iscore, position, kmer = [], [], []\n",
    "    for i in kmers['sequences']:\n",
    "        if '.' in i[core[0]:core[1]+1]: # added +1\n",
    "            iscore.append(False)\n",
    "        else:\n",
    "            iscore.append(True)\n",
    "        kPos = re.search('(C|G|A|T)',i).start() # 0-indexed\n",
    "        position.append(kPos) \n",
    "        kmer.append(i[kPos: kPos+kmer_len])\n",
    "        #kmer.append(re.findall('[A-Z]{' + str(kmer_length) + '}',i)[0])\n",
    "    kmers['is_core'] = iscore\n",
    "    kmers['kPosition'] = position\n",
    "    kmers['kmer'] = kmer # already filtered to be >= threshold?!!\n",
    "    return kmers\n",
    "\n",
    "def get_kmer_pos_OLD(peak_df, kmer_len, kmer_list): # add Item (aka pos in peak_df) as needed\n",
    "    # meant to combine kmer_match and score_blaster\n",
    "    # decide not to use b/c appraoch  unnecessarily convoluted\n",
    "    # replaced with get_kmer_pos()\n",
    "    '''\n",
    "    for each sequence within the peak_df, get the pos of kmers with E-score >= cutoff\n",
    "    '''\n",
    "    seq_list, pos_list = [], []\n",
    "    for row in zip(peak_df['seq'], peak_df['seq_id']):\n",
    "        seq, pos = row[0], row[1]\n",
    "        #print(seq, pos)\n",
    "        for i in range(len(seq) - kmer_len + 1):\n",
    "            window = seq[i:i+kmer_len]\n",
    "            \n",
    "            if window in kmer_list.values:\n",
    "                seq_list.append(window)\n",
    "                pos_list.append(i) # maybe add \"pos\" as well?\n",
    "    df = pd.DataFrame(list(zip(seq_list, pos_list)), columns =['Peak_Seq', 'Position']) # ['seq', 'pos']\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernel_info": {
   "name": "python3"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  },
  "nteract": {
   "version": "0.14.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
